backend/courses/management/commands/link_courses_to_topics.py

```diff
@@ -0,0 +1,215 @@
+import os
+from collections import defaultdict
+from textwrap import dedent
+
+import pandas as pd
+from django.core.management.base import BaseCommand
+from django.db.models import F, Max, OuterRef, Q, Subquery
+from tqdm import tqdm
+
+from alert.management.commands.export_anon_registrations import get_semesters
+from courses.management.commands.merge_topics import (
+    ShouldLinkCoursesResponse,
+    merge_topics,
+    should_link_courses,
+)
+from courses.models import Course, Topic
+from PennCourses.settings.base import S3_client
+
+
+def get_branches_from_cross_walk(cross_walk):
+    """
+    From a given crosswalk csv path, generate a dict mapping old_full_code to
+    a list of the new codes originating from that source, if there are multiple
+    (i.e. only in the case of branches).
+    """
+    branched_links = defaultdict(list)
+    cross_walk = pd.read_csv(cross_walk, delimiter="|", encoding="unicode_escape")
+    for _, r in cross_walk.iterrows():
+        old_full_code = f"{r['SRS_SUBJ_CODE']}-{r['SRS_COURSE_NUMBER']}"
+        new_full_code = f"{r['NGSS_SUBJECT']}-{r['NGSS_COURSE_NUMBER']}"
+        branched_links[old_full_code].append(new_full_code)
+    return {
+        old_code: new_codes for old_code, new_codes in branched_links.items() if len(new_codes) > 1
+    }
+
+
+def get_direct_backlinks_from_cross_walk(cross_walk):
+    """
+    From a given crosswalk csv path, generate a dict mapping new_full_code->old_full_code,
+    ignoring branched links in the crosswalk (a course splitting into multiple new courses).
+    """
+    guaranteed_links = dict()
+    cross_walk = pd.read_csv(cross_walk, delimiter="|", encoding="unicode_escape")
+    for _, r in cross_walk.iterrows():
+        old_full_code = f"{r['SRS_SUBJ_CODE']}-{r['SRS_COURSE_NUMBER']}"
+        new_full_code = f"{r['NGSS_SUBJECT']}-{r['NGSS_COURSE_NUMBER']}"
+        if old_full_code in guaranteed_links:
+            # Ignore branched links
+            del guaranteed_links[new_full_code]
```

Comment:
Could this lead to oscillatory behavior where if you have an even number of branches then there are no associations in the dict, but if you have an odd number of links there is an association?